pygmt/src/grdcut.py

```diff
@@ -99,13 +98,32 @@ def grdcut(grid, **kwargs):
     >>> # 12° E to 15° E and a latitude range of 21° N to 24° N
     >>> new_grid = pygmt.grdcut(grid=grid, region=[12, 15, 21, 24])
     """
-    with GMTTempFile(suffix=".nc") as tmpfile:
-        with Session() as lib:
-            with lib.virtualfile_in(check_kind="raster", data=grid) as vingrd:
-                if (outgrid := kwargs.get("G")) is None:
-                    kwargs["G"] = outgrid = tmpfile.name  # output to tmpfile
-                lib.call_module(
-                    module="grdcut", args=build_arg_list(kwargs, infile=vingrd)
-                )
+    # Determine the output data kind based on the input data kind.
+    inkind = data_kind(grid)
+    match inkind:
+        case "image" | "grid":
+            outkind = inkind
+        case "file":
+            realpath = str(grid)
+            if realpath.startswith("@"):  # Is a remote file without suffix
+                realpath = which(grid, download="a")
+                if isinstance(realpath, list):
+                    realpath = realpath[0]
 
-        return load_dataarray(outgrid) if outgrid == tmpfile.name else None
+            if realpath.endswith(
+                (".jpg", ".jpeg", ".png", ".tif", ".tiff", ".bmp", ".webp")
+            ):
```

Comment:
The list of image suffixes comes from https://github.com/GenericMappingTools/GMT.jl/blob/78a6f2d2ff2c2a00dd0b6760555fd6e02cf4484f/src/gmtreadwrite.jl#L10.

`jp2` is ignored here, because a `jp2` file can be either a grid or an image, which is complicated to support.

Reply:
If someone wants to write to a `.jp2` file, would it usually be a 1-band grid or a 3-band image? Can we not set `outkind=None` to let `inquire_virtualfile` determine this automatically based on the virtualfile?

Reply:
> If someone wants to write to a `.jp2` file, would it usually be a 1-band grid or a 3-band image?

Are you talking about **reading** a `.jp2` file? As I understand it, the [JPEG2000](https://en.wikipedia.org/wiki/JPEG_2000) format is designed for storing images, so it's usually 3-band, but in GMT, the tiled remote grids are stored in JPEG2000 format for its smaller size. These tiled grids are loaded like below. Is it a 1-band grid or a 1-band image?
```
>>> import rioxarray
>>> rioxarray.open_rasterio("S90W180.earth_age_05m_p.jp2")
<xarray.DataArray (band: 1, y: 2160, x: 2160)> Size: 9MB
[4665600 values with dtype=int16]
Coordinates:
  * band         (band) int64 8B 1
  * x            (x) float64 17kB -180.0 -179.9 -179.8 ... -0.125 -0.04167
  * y            (y) float64 17kB 89.96 89.88 89.79 ... -89.79 -89.87 -89.96
    spatial_ref  int64 8B 0
Attributes:
    scale_factor:  1.0
    add_offset:    0.0
```

> Can we not set `outkind=None` to let `inquire_virtualfile` determine this automatically based on the virtualfile?

`outkind` can be `None` in `virtualfile_to_raster`, but can't be `None` in `virtualfile_out`.

Reply:
No, I'm talking about writing a `.jp2` file. I'm assuming GMT can correctly read a `.jp2` file and infer whether it is a grid or image already.

> These tiled grids are loaded like below. Is it a 1-band grid or a 1-band image?

That `S90W180.earth_age_05m_p.jp2` file is supposed to be a 3-band image, but `rioxarray.open_rasterio` is incorrectly reading it as 1-band because rioxarray hasn't properly implemented `colormap`/`colorinterp` yet (tried at https://github.com/corteva/rioxarray/pull/414, but got a bit stuck).

> > Can we not set `outkind=None` to let `inquire_virtualfile` determine this automatically based on the virtualfile?
> 
> `outkind` can be `None` in `virtualfile_to_raster`, but can't be `None` in `virtualfile_out`.

Ah ok, so we do need to pick a kind here. From https://gdal.org/drivers/raster/jp2openjpeg.html, it says that jp2 supports arbitrary band counts, so either 3-band or 1-band is possible. I've seen both cases before in my past work, 3-band RGB images stored as JP2, and also 1-band images from Sentinel-2 (e.g. https://gis.stackexchange.com/questions/285887/converting-sentinel-2-jp2-bands-to-tiff-with-gdal), so a bit hesitant to pick one or the other :confused:

Reply:
Is it OK to say that a grid is equivalent to a 1-band image? Based on my local experiments, they seems to be the same. A one-band image (e.g., the `earth_day_01d_p.tif` file) can be read into either a `_GMT_GRID` container or a `_GMT_IMAGE` container.

Reply:
Will do some experiments in #3318. Will let you know if it works as expected.

Reply:
> Is it OK to say that a grid is equivalent to a 1-band image?

Yes, you can say that.

> Based on my local experiments, they seem to be the same. A one-band image (e.g., the `earth_day_01d_p.tif` file) can be read into either a `_GMT_GRID` container or a `_GMT_IMAGE` container.

Just to be clear, that `earth_day_01d_p.tif` file is not a 1-band image, it should be a 3-band image... Better check the pixel values if you're able to read it into a `_GMT_GRID`.